lib/spack/spack/util/module_cmd.py

```diff
@@ -24,6 +23,17 @@
 def module(*args, **kwargs):
     module_cmd = kwargs.get("module_template", "module " + " ".join(args))
 
+    # On ubuntu the default shell `/bin/sh`` is `dash`. As `bin/spack` uses `/bin/sh` there are no
+    # bash functions and hence no `module` command. We replace the `module` command by the actual
+    # function call in the bash function.
+    if not any(x.startswith("BASH_FUNC_module") for x in os.environ):
```

Comment:
Drop this then?

Reply:
Damn, this does not work for older versions. `MODULES_CMD` is only introduced in 4.1.

Reply:
The only consistent way I see is to ask the shell itself about what the module command looks like whenits invoked:
```
(set -x ; module foo ) 2>&1 | grep -- '^++' | grep foo | sed -e 's?^++\ ??g' -e 's?\ foo??g'
```
@haampie  do you think this  would be an option for the general case?

Reply:
uh, what?!

Reply:
I have a better version now. I use this to get the shell to tell me what it actually executes when calling "module ...":
```
(LMOD_SH_DBG_ON=1; set -x ; PS4=bar; module foo) 2>&1 | grep '^bbar.*foo' | sed -e 's?^.*bar??g' -e 's? foo??g'
```